cmd/swarm-smoke-pss/pss.go

```diff
@@ -0,0 +1,390 @@
+		if err != nil {
+			metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.fail", mode), nil).Inc(1)
+		}
+		return err
+	case <-time.After(time.Duration(timeout) * time.Second):
+		metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.timeout", mode), nil).Inc(1)
+		return fmt.Errorf("timeout after %v sec", timeout)
+	}
+
+}
+
+func pssSetup() *pssSession {
+	// random topic, one per session,  same for all msgs
+	topic := message.NewTopic(testutil.RandomBytes(seed, 4))
+
+	log.Trace("pss random topic", "topic", topic.String())
+
+	session := &pssSession{
+		msgC:  make(chan pss.APIMsg),
+		jobs:  make(map[string]*pssJob),
+		topic: topic,
+	}
+
+	// set up the necessary info for each pss node
+	for i, host := range hosts {
+		wsHost := fmt.Sprintf("ws://%s:%d", host, wsPort)
+
+		rpcClient, err := rpc.Dial(wsHost)
+		if err != nil {
+			log.Error("error dialing host", "err", err)
+			continue
+		}
+
+		// get overlay address
+		var addr hexutil.Bytes
+		err = rpcClient.Call(&addr, "pss_baseAddr")
+		if err != nil {
+			log.Error("error calling host for addr", "err", err)
+			continue
+		}
+
+		// get pss public key
+		var pubkey string
+		err = rpcClient.Call(&pubkey, "pss_getPublicKey")
+		if err != nil {
+			log.Error("error calling host for pubkey", "err", err)
+			continue
+		}
+
+		// subscribe to the topic for the session
+		// this creates the incoming message handler automatically
+		// any message received with this topic comes on the channel
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		sub, err := rpcClient.Subscribe(ctx, "pss", session.msgC, "receive", session.topic, true, false)
+		if err != nil {
+			log.Error("error calling host for subscribe", "err", err)
+			continue
+		}
+
+		var dFn = func() {
+			sub.Unsubscribe()
+			rpcClient.Close()
+		}
+		session.nodes = append(session.nodes,
+			&pssNode{
+				hostIdx:        i,
+				addr:           []byte(addr),
+				pubkey:         pubkey,
+				peerPubKeys:    make(map[string]bool),
+				client:         rpcClient,
+				deregisterFunc: dFn,
+			},
+		)
+	}
+	return session
+}
+
+// processJobs processes the given jobs and returns the success count and an error
+func (s *pssSession) processJobs(jobs []pssJob) (int, error) {
+	errs := []error{}
+	for _, j := range jobs {
+		var err error
+		switch mode := j.mode; mode {
+		case pssModeRaw:
+			err = s.sendRawMessage(j.sender, j.receiver, j.msg)
+			if err != nil {
+				log.Error("error sending raw message", "err", err)
+			}
+		case pssModeSym:
+			err = s.sendSymMessage(j.sender, j.receiver, j.msg)
+			if err != nil {
+				log.Error("error sending sym message", "err", err)
+			}
+		case pssModeAsym:
+			err = s.sendAsymMessage(j.sender, j.receiver, j.msg)
+			if err != nil {
+				log.Error("error sending asym message", "err", err)
+			}
+		default:
+			err = fmt.Errorf("invalid pssMode %d", mode)
+			log.Error("error processing job type", "err", err)
+		}
+
+		err = s.waitForMsg()
```

Comment:
I assume the only thing we want to check here is correctness in the sense that messages are actually delivered, regardless of performance?

Reply:
let's start with that, yes